### PR TITLE
[READY] Update SuperChart onRenderXXX listeners

### PR DESCRIPTION
**Pull request by kristw**
This PR will improve the accuracy of rendering time logging and also resolve the issue due to out-of-react-cycle execution. 

- Remove `skipRendering` flag. Check for existence of `chartProps` instead. 
- Dispatch listener `onRenderXXX` from the correct place. Before, I was dispatching this from `SuperChart` which does not have full transparency to when the rendering actually happens because the rendering happens in a nested component created by `react-loadable`. This PR extends the nested component generated by `react-loadable` to provide listeners after rendering and use that to trigger `onRenderSuccess` and `onRenderFailure` more accurately.
- Manually trigger `preloading` of lazy-loaded vis code right after `SuperChart` knows the `chartType`. The previous code won't load until `chartProps` is available

@mistercrunch @williaster  @graceguo-supercat @michellethomas @xtinec @conglei

### CODE REVIEW

File: superset/assets/src/visualizations/core/components/SuperChart.jsx [hunk 1]
```diff
@@ -130,14 +161,18 @@ class SuperChart extends React.PureComponent {
...
+    // Create LoadableRenderer and start preloading
+    // the lazy-loaded Chart components
     const LoadableRenderer = this.createLoadableRenderer(this.props);
 
-    // Use this to allow loading the vis components
-    // without rendering (while waiting for data)
-    if (skipRendering || !chartProps) {
+    // Do not render if chartProps is not available.
+    // but the pre-loading has been started in this.createLoadableRenderer
+    // to prepare for rendering once chartProps becomes available.
```
Comment (williaster) [inline superset/assets/src/visualizations/core/components/SuperChart.jsx]: 👍

**Review by williaster — COMMENTED** [review]
Looks great overall, I had one suggestion about splitting the custom renderer as it's a little hard to follow.

File: superset/assets/src/visualizations/core/components/SuperChart.jsx [hunk 2]
```diff
@@ -76,12 +74,49 @@ class SuperChart extends React.PureComponent {
...
             render: this.renderChart,
           });
+
+          // Trigger preloading.
+          LoadableRenderer.preload();
+
+          // Extends the behavior of LoadableComponent
+          // generated by react-loadable
+          // to provide post-render listeners
+          class CustomLoadableRenderer extends LoadableRenderer {
```
[the remark this answers is not in the document]
Comment (kristw) [reply superset/assets/src/visualizations/core/components/SuperChart.jsx]: Yeah. Wish they have a post-load hook. This is a bit fragile if they change how it is written. 

Anyway, good call on separating into another factory file. Will do that.

**Comment by kristw** [comment]
Updated and ready for review

File: superset/assets/spec/javascripts/visualizations/core/createLoadableRenderer_spec.jsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      shallow(
+        <LoadableRenderer
+          onRenderSuccess={onRenderSuccess}
+          onRenderFailure={onRenderFailure}
+        />,
+      );
+      expect(loadChartSuccess).toHaveBeenCalled();
+      setTimeout(() => {
+        expect(render).toHaveBeenCalled();
+        expect(onRenderSuccess).toHaveBeenCalled();
```
Comment (williaster) [inline superset/assets/spec/javascripts/visualizations/core/createLoadableRenderer_spec.jsx]: would it be useful to add a specific `callCount=1` in these tests? `CustomLoadableRenderer` calls `afterRender` in `componentDidMount` and `componentDidUpdate` so just want to make sure it doesn't call it 2x.

File: superset/assets/src/visualizations/core/components/createLoadableRenderer.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+};
+
+export default function createLoadableRenderer(options) {
+  const LoadableRenderer = Loadable.Map(options);
+
+  // Extends the behavior of LoadableComponent
+  // generated by react-loadable
+  // to provide post-render listeners
+  class CustomLoadableRenderer extends LoadableRenderer {
+    componentDidMount() {
```
Comment (williaster) [inline superset/assets/src/visualizations/core/components/createLoadableRenderer.js]: related to my other comment, but what's the reasoning for adding this in `mount` + `update` instead of just `update`?

**Review by williaster — APPROVED** [review]
this looks great! much easier to follow in separate files I think 👍 

had a question about the `afterRender` calls but pending that I think it's good to go!

File: superset/assets/spec/javascripts/visualizations/core/createLoadableRenderer_spec.jsx (hunk 3 above)
Comment (kristw) [reply superset/assets/spec/javascripts/visualizations/core/createLoadableRenderer_spec.jsx]: 👍

File: superset/assets/src/visualizations/core/components/createLoadableRenderer.js (hunk 4 above)
Comment (kristw) [reply superset/assets/src/visualizations/core/components/createLoadableRenderer.js]: `componentDidUpdate` will be called after 2nd, 3rd, ... `render`
but after the first `render` will only call `componentDidMount`.
